### PR TITLE
ast-grep: 11.1 -> 12.2

### PR DESCRIPTION
**Pull request by cafkafk**
## Description of changes
The [`test_unmatching_id`](https://github.com/ast-grep/ast-grep/blob/9c78663c100123b4cc33b394e08f200cf6ade59f/crates/cli/src/verify/test_case.rs#L190) test broke for `v0.12.1`. I was wondering if we could ignore this.

My current plan was to add this as a separate commit, mostly to make it explicit that ast-grep ignores that test for reviewers, but if you'd prefer having it in the same commit, I could do that instead.

Changelog: https://github.com/ast-grep/ast-grep/releases

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by cafkafk** [comment]
This still seems to fail with 4 more broken tests. This is probably something to be solved upstream.

**Comment by cafkafk** [comment]
I'll update it now, upstream fixed it

**Comment by cafkafk** [comment]
I still get the error on the new version:
```
ast-grep> test verify::test_case::tests::test_verify_snapshot_with_existing ... ok
ast-grep> test verify::test_case::tests::test_verify_snapshot_without_existing_2 ... ok
ast-grep> test print::interactive_print::test::test_open_editor ... ok
ast-grep> test test_cli::test_run ... ok
ast-grep> ./src/verify/mod.rs:88:      Some(result)
ast-grep> test run::test::test_run_with_specific_lang ... ok
ast-grep> test scan::test::test_run_with_config ... ok
ast-grep> test run::test::test_run_with_pattern ... ok
ast-grep> failures:
ast-grep> ---- verify::test_case::tests::test_unmatching_id stdout ----
ast-grep> note: test did not panic as expected
ast-grep> failures:
ast-grep>     verify::test_case::tests::test_unmatching_id
ast-grep> test result: FAILED. 64 passed; 1 failed; 1 ignored; 0 measured; 0 filtered out; finished in 0.04s
ast-grep> error: test failed, to rerun pass `-p ast-grep --lib`
```

**Comment by Lord-Valen** [comment]
That one seems to be unique to the nix build. I can't reproduce it in the source repo and per my reading of the code, it *should* panic.

**Comment by cafkafk** [comment]
> That one seems to be unique to the nix build. I can't reproduce it in the source repo and per my reading of the code, it _should_ panic.

I'm gonna try to build it without it, hopefully it works now.

**Comment by cafkafk** [comment]
This actually build fine for me locally now... I really hope ofborg also likes it :pray:

**Comment by cafkafk** [comment]
It had two failing tests on `aarch64-linux` and `x86_64-darwin`, I'm just gonna ignore them, if ofborg likes this maybe someone should try this out on a MacOS computer and see if it's broken.

**Review by dundargoc — APPROVED** [review]

File: pkgs/by-name/as/ast-grep/package.nix [hunk 1]
```diff
@@ -25,6 +25,13 @@ rustPlatform.buildRustPackage rec {
   checkFlags = [
     # disable flaky test
     "--skip=test::test_load_parser_mac"
+
+    # BUG: Broke by 0.12.1 update (https://github.com/NixOS/nixpkgs/pull/256548)
+    # Please check if this is fixed in future updates of the package
+    "--skip=verify::test_case::tests::test_unmatching_id"
+  ] ++ lib.optionals (stdenv.isDarwin || stdenv.isAarch64) [
+    "--skip=test::test_load_parser"
+    "--skip=test::test_register_lang"
```
Comment (Artturin) [inline pkgs/by-name/as/ast-grep/package.nix:34]:
```suggestion
    # x86_64-darwin: source/benches/fixtures/json-mac.so\' (no such file), \'/private/tmp/nix-build-.../source/benches/fixtures/json-mac.so\' (mach-o file, but is an incompatible architecture (have \'arm64\', need \'x86_64h\' or \'x86_64\'))" })
    # aarch64-linux: /build/source/benches/fixtures/json-linux.so: cannot open shared object file: No such file or directory"
    "--skip=test::test_load_parser"
    "--skip=test::test_register_lang"
```

File: pkgs/by-name/as/ast-grep/package.nix [hunk 2]
```diff
@@ -25,6 +25,13 @@ rustPlatform.buildRustPackage rec {
   checkFlags = [
     # disable flaky test
     "--skip=test::test_load_parser_mac"
+
+    # BUG: Broke by 0.12.1 update (https://github.com/NixOS/nixpkgs/pull/256548)
+    # Please check if this is fixed in future updates of the package
+    "--skip=verify::test_case::tests::test_unmatching_id"
+  ] ++ lib.optionals (stdenv.isDarwin || stdenv.isAarch64) [
```
Comment (Artturin) [inline pkgs/by-name/as/ast-grep/package.nix:32]:
```suggestion
  ] ++ lib.optionals ((stdenv.isDarwin && stdenv.isAarch64) || (stdenv.isLinux && stdenv.isAarch64)) [
```

File: pkgs/by-name/as/ast-grep/package.nix [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    # BUG: Broke by 0.12.1 update (https://github.com/NixOS/nixpkgs/pull/257385)
+    # Please check if this is fixed in future updates of the package
+    "--skip=verify::test_case::tests::test_unmatching_id"
+  ] ++ lib.optionals (with stdenv.hostPlatform; (isDarwin && isx86_64) || (isLinux && isAarch64)) [
+    # x86_64-darwin: source/benches/fixtures/json-mac.so\' (no such file), \'/private/tmp/nix-build-.../source/benches/fixtures/json-mac.so\' (mach-o file, but is an incompatible architecture (have \'arm64\', need \'x86_64h\' or \'x86_64\'))" })
+    # aarch64-linux: /build/source/benches/fixtures/json-linux.so: cannot open shared object file: No such file or directory"
+    "--skip=test::test_load_parser"
+    "--skip=test::test_register_lang"
+  ];
+
```
Comment (Artturin) [inline pkgs/by-name/as/ast-grep/package.nix:38]: I recommend adding a `passthru.tests.version`

Comment (cafkafk) [reply pkgs/by-name/as/ast-grep/package.nix:38]: I've never done that before, but I could try?

**Pull request merged**